File: mobile/modules/expo-pitch-detector/android/src/main/java/expo/modules/pitchdetector/ExpoPitchDetectorModule.kt
// modules/expo-pitch-detector/android/src/main/java/expo/modules/pitchdetector/ExpoPitchDetectorModule.kt
package expo.modules.pitchdetector

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoPitchDetectorModule : Module() {
  private var capture: AudioCapture? = null
  // Deliberately still 22050 while iOS runs at 48000: `windowSize` and OnsetDetector's
  // hop are sample counts, so raising the rate here without doubling both would halve
  // the analysis window and cost the low strings their NSDF reliability. Raise all
  // three together, or not at all — and note that a `sampleRate` passed from JS reaches
  // both platforms, so it would hit exactly that trap.
  private var sampleRate: Int = 22050
  private val windowSize = 2048
  private val hop = 1024
  private var detector = PitchDetectorMPM(sampleRate.toDouble(), windowSize)
  private val ring = ArrayDeque<Float>()
  private val ringLock = Any()

  // Outlives any one capture so `configureOnsets` means something before start() and
  // survives stop(). It carries its own lock, so it is reachable from every thread.
  private val onsets = OnsetDetector()

  override fun definition() = ModuleDefinition {
    Name("ExpoPitchDetector")
    Events("onPitch", "onOnset")

    AsyncFunction("configureOnsets") { config: Map<String, Any> ->
      onsets.configure(
        enabled = (config["enabled"] as? Boolean) ?: false,
        threshold = (config["threshold"] as? Number)?.toFloat() ?: 0f,
        refractoryMs = (config["refractoryMs"] as? Number)?.toDouble() ?: 0.0,
      )
    }

    AsyncFunction("start") { opts: Map<String, Any>? ->
      val ctx = appContext.reactContext
        ?: throw IllegalStateException("PERMISSION_DENIED")
      val granted = androidx.core.content.ContextCompat.checkSelfPermission(
        ctx, android.Manifest.permission.RECORD_AUDIO,
      ) == android.content.pm.PackageManager.PERMISSION_GRANTED
      if (!granted) throw IllegalStateException("PERMISSION_DENIED")

      // Idempotent: stop any prior capture before reconfiguring.
      capture?.stop()
      capture = null

      opts?.get("sampleRate")?.let { sampleRate = (it as Number).toInt() }
      detector = PitchDetectorMPM(sampleRate.toDouble(), windowSize)
      synchronized(ringLock) { ring.clear() }
      onsets.begin(sampleRate.toDouble())

      val cap = AudioCapture(sampleRate) { samples, startEpochMs -> handle(samples, startEpochMs) }
      cap.start()
      capture = cap
    }

    AsyncFunction("stop") {
      capture?.stop()
      capture = null
      synchronized(ringLock) { ring.clear() }
      // Config survives; only the envelope history and any undrained detections go.
      onsets.reset()
    }

    OnDestroy {
      capture?.stop()
      capture = null
      onsets.reset()
    }
  }

  private fun handle(samples: FloatArray, startEpochMs: Double) {
    // Onsets first: their timestamps are their own, but the MPM pass below is the slow
    // part of this block and there is no reason to sit behind it.
    onsets.feed(samples, startEpochMs)
    for (onset in onsets.drain()) {
      sendEvent("onOnset", mapOf("at" to onset.atMs, "peak" to onset.peak.toDouble()))
    }

    // Drain into a local window buffer under lock, then run MPM and emit outside the lock
    // to keep the critical section short.
    val windows = mutableListOf<FloatArray>()
    synchronized(ringLock) {
      for (s in samples) ring.addLast(s)
      while (ring.size >= windowSize) {
        val win = FloatArray(windowSize) { ring.elementAt(it) }
        repeat(hop) { ring.removeFirst() }
        windows.add(win)
      }
    }
    for (win in windows) {
      val rms = detector.rms(win)
      val r = detector.detect(win)
      sendEvent(
        "onPitch",
        mapOf(
          "frequency" to (r?.frequency ?: 0.0),
          "clarity" to (r?.clarity ?: 0.0),
          "rms" to rms,
          "timestamp" to System.currentTimeMillis().toDouble(),
        ),
      )
    }
  }
}

File: mobile/modules/expo-pitch-detector/ios/ExpoPitchDetectorModule.swift
import AVFoundation
import Accelerate
import ExpoModulesCore

public class ExpoPitchDetectorModule: Module {
  // 48kHz is the hardware rate on every modern iPhone, so the converter in AudioEngine
  // passes samples through instead of resampling. The window is sized in *time*, not
  // samples: NSDF needs several periods of the lowest note in view, so halving the
  // window duration to keep 2048 samples would cost the low strings their reliability.
  // 4096 @ 48kHz = 85ms, which holds ~4.7 periods of the 55Hz `minFreq` floor.
  private let windowSize = 4096
  private let defaultSampleRate: Double = 48000

  // engine/ring/detector/timer are owned by processingQueue: they are only read or written
  // from it (in tick, startProcessing, and the publish/teardown blocks). The real-time audio
  // tap never touches them — it writes into a ring instance captured directly in its closure.
  private let processingQueue = DispatchQueue(label: "expo.pitchdetector.processing")
  private var engine: AudioEngine?
  private var ring: SampleRing?
  private var detector: PitchDetectorMPM?
  private var timer: DispatchSourceTimer?

  // Outlives any one session so `configureOnsets` means something before start() and
  // survives stop(). It carries its own lock, so it is reachable from every queue.
  private let onsets = OnsetDetector()

  public func definition() -> ModuleDefinition {
    Name("ExpoPitchDetector")

    Events("onPitch", "onOnset")

    AsyncFunction("configureOnsets") { (config: [String: Any]) -> Void in
      self.onsets.configure(
        enabled: (config["enabled"] as? Bool) ?? false,
        threshold: Float((config["threshold"] as? Double) ?? 0),
        refractoryMs: (config["refractoryMs"] as? Double) ?? 0
      )
    }

    AsyncFunction("start") { (opts: [String: Any]?) async throws -> Void in
      // Idempotent: tear down any prior session before reconfiguring.
      self.teardown()

      let status = AVCaptureDevice.authorizationStatus(for: .audio)
      switch status {
      case .authorized:
        break
      case .notDetermined:
        let granted = await AVCaptureDevice.requestAccess(for: .audio)
        if !granted {
          throw NSError(domain: "ExpoPitchDetector", code: 100, userInfo: [
            NSLocalizedDescriptionKey: "PERMISSION_DENIED",
          ])
        }
      default:
        throw NSError(domain: "ExpoPitchDetector", code: 100, userInfo: [
          NSLocalizedDescriptionKey: "PERMISSION_DENIED",
        ])
      }

      let sr = (opts?["sampleRate"] as? Double) ?? defaultSampleRate
      let ring = SampleRing(capacity: self.windowSize)
      let detector = PitchDetectorMPM(sampleRate: sr, windowSize: self.windowSize)

      let onsets = self.onsets
      onsets.begin(sampleRate: sr)

      // The tap captures these instances directly, so the real-time audio thread never
      // reads a shared mutable property.
      let engine = AudioEngine(sampleRate: sr) { samples, count, startEpochMs in
        ring.write(samples, count: count)
        // Fed from the converter output, not from the ring: hops must be consecutive and
        // the ring only keeps the freshest window.
        onsets.feed(samples, count: count, startEpochMs: startEpochMs)
      }
      try engine.start()

      // Publish the session on processingQueue so it is sequenced with tick()/teardown().
      self.processingQueue.async {
        self.engine = engine
        self.ring = ring
        self.detector = detector
        self.startProcessing()
      }
    }

    AsyncFunction("stop") { () -> Void in
      self.teardown()
    }

    OnDestroy {
      self.teardown()
    }
  }

  // Must be called on processingQueue.
  private func startProcessing() {
    let t = DispatchSource.makeTimerSource(queue: processingQueue)
    // 15ms rather than 30 because the JS filter chain downstream (Median3, then the EMA
    // in TunerGate) is measured in *frames*, not milliseconds — halving the period
    // halves their wall-clock latency too, which is most of what this buys. Leeway
    // shrinks with it; at 15ms a 5ms slip is a third of the period.
    t.schedule(deadline: .now() + .milliseconds(15),
               repeating: .milliseconds(15),
               leeway: .milliseconds(2))
    t.setEventHandler { [weak self] in self?.tick() }
    self.timer = t
    t.resume()
  }

  // Runs on processingQueue at ~15ms cadence. Always analyzes the freshest window; the timer
  // coalesces missed firings, so a slow tick can't pile up a backlog of catch-up work.
  private func tick() {
    // Drained ahead of the pitch guard below: onsets must not wait on the ring filling,
    // and their timestamps are their own — a late tick does not make them late.
    for onset in onsets.drain() {
      sendEvent("onOnset", [
        "at":   onset.atMs,
        "peak": Double(onset.peak),
      ])
    }

    guard let ring = ring, let detector = detector, let window = ring.snapshot() else { return }

    var meanSquare: Float = 0
    vDSP_measqv(window, 1, &meanSquare, vDSP_Length(window.count))
    let rms = sqrt(meanSquare)

    let result = detector.detect(window)
    sendEvent("onPitch", [
      "frequency": result?.frequency ?? 0,
      "clarity":   result?.clarity   ?? 0,
      "rms":       Double(rms),
      "timestamp": Date().timeIntervalSince1970 * 1000,
    ])
  }

  private func teardown() {
    processingQueue.sync {
      timer?.cancel()
      timer = nil
      engine?.stop()
      engine = nil
      ring = nil
      detector = nil
      // Config survives; only the envelope history and any undrained detections go.
      onsets.reset()
    }
  }
}
